go-controller/pkg/clustermanager/egressip_controller.go

```diff
@@ -0,0 +1,1768 @@
+		}
+		mgmtIPs := make([]net.IP, len(nodeSubnets))
+		for i, subnet := range nodeSubnets {
+			mgmtIPs[i] = util.GetNodeManagementIfAddr(subnet).IP
+		}
+		eIPC.allocator.cache[node.Name] = &egressNode{
+			name:           node.Name,
+			egressIPConfig: parsedEgressIPConfig,
+			mgmtIPs:        mgmtIPs,
+			allocations:    make(map[string]string),
+			healthClient:   hccAllocator.allocate(node.Name),
+		}
+	}
+	return nil
+}
+
+// deleteAllocatorEgressIPAssignments deletes the allocation as to keep the
+// cache state correct, also see addAllocatorEgressIPAssignments
+func (eIPC *egressIPController) deleteAllocatorEgressIPAssignments(statusAssignments []egressipv1.EgressIPStatusItem) {
+	eIPC.allocator.Lock()
+	defer eIPC.allocator.Unlock()
+	for _, status := range statusAssignments {
+		if eNode, exists := eIPC.allocator.cache[status.Node]; exists {
+			delete(eNode.allocations, status.EgressIP)
+		}
+	}
+}
+
+// deleteAllocatorEgressIPAssignmentIfExists deletes egressIP config from node allocations map
+// if the entry is available and returns assigned node name, otherwise returns empty string.
+func (eIPC *egressIPController) deleteAllocatorEgressIPAssignmentIfExists(name, egressIP string) string {
+	eIPC.allocator.Lock()
+	defer eIPC.allocator.Unlock()
+	for nodeName, eNode := range eIPC.allocator.cache {
+		if egressIPName, exists := eNode.allocations[egressIP]; exists && egressIPName == name {
+			delete(eNode.allocations, egressIP)
+			return nodeName
+		}
+	}
+	return ""
+}
+
+// addAllocatorEgressIPAssignments adds the allocation to the cache, so that
+// they are tracked during the life-cycle of ovnkube-master
+func (eIPC *egressIPController) addAllocatorEgressIPAssignments(name string, statusAssignments []egressipv1.EgressIPStatusItem) {
+	eIPC.allocator.Lock()
+	defer eIPC.allocator.Unlock()
+	for _, status := range statusAssignments {
+		if eNode, exists := eIPC.allocator.cache[status.Node]; exists {
+			eNode.allocations[status.EgressIP] = name
+		}
+	}
+}
+
+func (eIPC *egressIPController) reconcileEgressIP(old, new *egressipv1.EgressIP) (err error) {
+	// Lock the assignment, this is needed because this function can end up
+	// being called from WatchEgressNodes and WatchEgressIP, i.e: two different
+	// go-routines and we need to make sure the assignment is safe.
+	eIPC.egressIPAssignmentMutex.Lock()
+	defer eIPC.egressIPAssignmentMutex.Unlock()
+
+	// Initialize an empty name which is filled depending on the operation
+	// (ADD/UPDATE/DELETE) we are performing. This is done as to be able to
+	// delete the NB DB set up correctly when searching the DB based on the
+	// name.
+	name := ""
+
+	// Initialize a status which will be used to compare against
+	// new.spec.egressIPs and decide on what from the status should get deleted
+	// or kept.
+	status := []egressipv1.EgressIPStatusItem{}
+
+	// Initialize two empty objects as to avoid SIGSEGV. The code should play
```

Comment:
NIT: I find this comment confusing, there is only one object initialized here?

Reply:
yea thanks again, i SHOULD have read each line.. and ensured comments didn't go stale. let me try and do that at least across the reconcile function.